rfc: wing cloud preview environments (#3696)
winglang/wing: docs/docs/09-rfcs/2023-03-08-wing-preview-environment.md

---
title: Wing Cloud Preview Environments
description: This document describes the Wing Cloud Preview Environments feature.
---

# #3696 - Wing Cloud Preview Environments

- **Author(s)**: @ainvoner, @skyrpex
- **Submission Date**: 2023-08-03
- **Stage**: Proposal
- **Stage Date**: 2023-08-03

This document describes the Wing Cloud Preview Environments feature.

## Introduction
The **Wing Cloud Preview Environments** project is designed to achieve two key objectives. Firstly, it aims to offer a preview environment solution for those interested in using winglang.
Secondly, it seeks to enhance the winglang ecosystem and its toolchain by building a production application using the language.

## Requirements
Create the first production feature of Wing Cloud - Wing cloud Preview Environments.
Build and Deploy wing applications to a Wing Console simulated environment.

> Allow developers to iterate with their teammates using a shared Wing Console instance.

Author: updated the text, the reason for adding it as a requirement is to make sure the environment and data will be shared.

Seamlessly integrating with GitHub.

### Developer Experience
A developer using winglang has the capability to integrate preview environments with his/her repository.

#### Sign Up and Installation
In order to use Wing Cloud Preview Environments, you should sign up to Wing Cloud.
The signup and installation process is straightforward using GitHub application:

> 1. Goto https://wing.cloud website and click "Deploy with Wing Cloud"

Author: why login / signup? we would like them to deploy their application, during this process they signup to wing cloud, authenticate via GitHub and grant access to a repository

> 1. Goto https://wing.cloud website and click "Deploy with Wing Cloud"

Reviewer: vercel's CTA is "start deploying" (which leads to https://vercel.com/new but they also have a "login" and "sign up" buttons on their home page.

let's do the same thing.

> 1. Goto https://wing.cloud website and click "Deploy with Wing Cloud"

Author: I came to a conclusion that we can't avoid a sign up process to Wing Cloud which should not be dependent on the Preview Environment feature.
I think we need a different issue to discuss and define it

2. Complete GitHub authentication.
3. Grant repository access permission.
4. You are now signed up and ready to use Wing Cloud Preview Environments.

#### Preview Environment for production branch

> For the production branch, there is a single preview environment that will constantly be up-to-date with the latest code committed.

Reviewer: how will URLs look like?

Ideally something like:

```
https://wing.cloud/gh-account/gh-repo/gh-branch
```

e.g.

https://wing.cloud/winglang/docsite/main

---

this should be the same for PRs (because they must be branched), no?

Author: I thought of using subdomains for wing.cloud:
1. Production branch environment url structure is: `https://<gh-repository>-<production-branch-name>.wing.cloud.app`
2. PR preview environments urls structure is: `https://<gh-repository>-<gh-branch>-<pr-number>-<entry-point>.wing.cloud.app`

Reviewer: Why not normal API routes? What's the benefit of using subdomains? I think normal URL routes are more suitable for this need.

Reviewer 2: Using subdomains is ~preferred~ mandatory here, security wise.

The preview environments will run code created by users. If we serve them under wing.cloud, we expose us to [XSRF attacks](https://en.wikipedia.org/wiki/Cross-site_request_forgery).

Even if this preview environment is served under the Console App control, it's very likely that users may be able to abuse it. Attacks that would be possible are:

- Gathering information from the user previewing the environment
- Executing calls to wing.cloud in the name of this user

I think we can run the subdomains under `*.wing.cloud` but I need to do some research first... there's some HTTP headers that can protect the main site from the subdomains. But the safest option is to use a totally different domain, something like `*.wing-preview.cloud`, so there's no open doors for XSRF.

Author: @skyrpex isn't wing.cloud.app is a different domain than wing.cloud? isn't it enough?

Reviewer: I am okay to use subdomains if that's the only way to defend against attacks but it should be under the wing.cloud parent

Author: but we need a different domain... accessing our wing.cloud auth cookie for example would be possible from preview environments if we will use sub domains of wing.cloud.

Reviewer 2: > I am okay to use subdomains if that's the only way to defend against attacks but it should be under the wing.cloud parent

I'm afraid we'll have to use a different domain altogether. There's multiple security risks involved...

Just to give you an example:
- After signing in to `https://wing.cloud`, there'll be an secure HTTP-only cookie on `wing.cloud` used for authentication
- If the preview environments are `*.wing.cloud`, the browser will send this cookie to them and they'll be able to read your session credentials and stuff on your behalf

There are security measures we could take to prevent attacks, but it'd be like if the whole Wing Cloud's security was hanging from a thread. The smallest mistake would become a security breach. We need to go for something like `*.wingcloud.app`.

A link to the preview environment is available in the repository main page ("About" section).

#### Preview Environments for Pull Requests

> Upon each creation of a pull request, an automatic comment will be added to the PR and will guide the developer to a dedicated preview environment. The deployed preview environment will have a unique url for easy identification.

Reviewer: see above re urls

To provide a streamlined process for updating a pull request preview environment, each PR code changes will redeploy the preview environment and also will:
1. provide a real-time build and deployment updates
2. each entry point will have a seperated preview environment
3. ensure a consistent url for each preview environment
4. provide the ability to download the preview environment deployment logs for debugging purposes in case of failure

> 

Reviewer: what happens if the branch is private? how do we ensure other people can't access it? do we use github auth there too?

Author: perhaps for our MVP we can assume that all users allowed to view the PR can also view the preview env. and logs.
what do you think?

Reviewer: yes, but just bear in mind that this means that access to the environment must be authenticated and there should be some authorization performed against github.

Author: accessing env? or creating new ones? even vercel doesn't require authentication for private repositories preview environments.
We had to pay for this service when our console was a private repo.
I don't think we should limit viewing environments.

Reviewer 2: The way Vercel works is what makes most sense IMO:

- GitHub authentication is only used to authorize the Wing Cloud GitHub App to have access to the repository
- The Preview Environment can be private but in order to see it, a user needs to be logged in to Wing Cloud and have view permission to the environment (the permissions are controlled by Wing Cloud, not GitHub anymore)

This is why is makes sense to me:

- A GitHub user having read permission to the repository code doesn't mean they should be able to see the preview environment
- Our auth code won't need to closely work along with GitHub's API, so it'll be much easier to maintain

PR comment example:

| Entry Point     | Status | Preview | Updated (UTC)        |
|-----------------| ------------- | ------------- |----------------------| 

> | main.w          |  ✅ Ready ([logs](https:/main-pr.wing-preview.com/logs)) | [Visit Preview](https:/main-pr.wing-preview.com) | Jul 31, 2023 8:01am  |

Reviewer: fix urls

| another.main.w  |  ✅ Ready ([logs](https:/main-pr.wing-preview.com/logs)) | [Visit Preview](https:/main-pr.wing-preview.com) | Jul 30, 2023 10:01am |
| failed.main.w   | ❌ Failed ([logs](https:/main-pr.wing-preview.com/logs))  |  | Jul 31, 2023 8:01am  |
| building.main.w |  🔄 Building ([logs](https:/main-pr.wing-preview.com/logs))  | | Jul 31, 2023 8:01am  |
| stale.main.w    | ⏸️ Stale  ([lean more](https:/winglang.io/docs/preview))  | | Jul 31, 2023 8:01am  |

> #### Multiple Entry Points

Author: the motivation to support a configuration file is to support our own dog fooding and to run wing console preview environment.
we will need to run the locally built wing to test the console new code

> #### Multiple Entry Points

Reviewer: Let's be careful with introducing config files. If we do, these should be read by the Wing CLI and not by the preview environments system.

* Entryfile(s) should be any file called `main.w` or has the `.main.w` extension.
* Install is always through `npm` right now.
* What's `buildScript`?

> #### Multiple Entry Points

Reviewer 2: What if we go (again) the Vercel route?

- In Wing Cloud, you can create more than one Preview Environment for a GitHub repository
- When you create a new Preview Environment, you select the repository
- Then, you select the wing entry file, and maybe some more configuration
- The configuration stays on Wing Cloud, so we don't need to define a config file

> #### Multiple Entry Points

Reviewer: Building websites is something we should discuss outside of the scope of preview environments. It's relevant for local development as well. Let's open an issue and figure this out separately. Preview environments will use the same mechanism.

Regarding wing version, we *will* have some form of config for Wing, but again, this is not a preview environment concern. Let's open an issue and discuss separately.

A developer can configure multiple entry points in a single PR.
Every `*.main.w` file in the repository will be considered as an entry point and will have a dedicated preview environment.
All preview environments will be available in the PR comment and will be updated upon code changes.

#### Populate Environment with Initial Data
TBD

#### Running Tests

> Running tests on your environment is easy using the Wing Console tests UI.

Reviewer: i think we should also run tests automatically and report to the PR

#### Self-Cleaning Environments Mechanism
To ensure efficient resource utilization:
1. Preview environments associated with stale PRs (without access or code changes for over 30 days) will automatically deactivate.
2. The PR's preview environment comment will indicate its inactive status.
3. Changes in code will trigger redeployment for all related preview environments in this PR.

### Analytics and Logs
To improve the product we are collecting data and monitoring the system.
1. We collect preview environments usage analytics only. We won't collect anything about your code and project.
2. OpenTelemetry is used for collecting metrics, logs, traces.
3. Datadog is used for monitoring the system.

### Documentation
Developers reading our docs can easily sign up to Wing Cloud and start using its Preview Environments feature in their repository.
In our docs we have the following information:
1. What is Wing Cloud Preview Environments
2. Getting started instructions
3. How can I configure multiple entry points
4. Populate environment with initial data
5. Running tests

### Security
1. The preview environments shouldn’t be able to access files from other preview environments

> 2. Only authenticated users should be able to use Wing Preview Environments

Reviewer: If a preview environment is working against a private GitHub repository, I think we will need to make sure that only GitHub users that have access to this repo can view the environment, no?

> 2. Only authenticated users should be able to use Wing Preview Environments

Author: see https://github.com/winglang/wing/pull/3696#discussion_r1285576005
I don't think we should handle it at the moment

> 2. Only authenticated users should be able to use Wing Preview Environments

Reviewer 2: I'd implement a USER_x_PROJECT permission on our end, as stated here: https://github.com/winglang/wing/pull/3696#discussion_r1285605394

> 2. Only authenticated users should be able to use Wing Preview Environments

Author: I think we should start with a simple gh auth for wing cloud without any extra permissions.
not all signed in users to wing cloud would like to install the preview app in their repo.

2. Goto https://wing.cloud/preview and Click on "Deploy with Wing Cloud"
3. Complete GitHub application authentication.
4. Grant repository access permission.
5. You are now ready to use Wing Cloud Preview Environments.
6. A welcome email will be sent to you with additional information and links to documentation.

#### Preview Environment for production branch

For the production branch, there is a single preview environment that will constantly be up-to-date with the latest code committed.
A link to the preview environment is available in the repository main page ("About" section).
Production branch environment url structure is: `https://<gh-repository>-<gh-branch>-<entry-point>.wing.cloud.app`

#### Preview Environments for Pull Requests

Upon each creation of a pull request an automatic comment will be added to the PR and will guide the developer to dedicated preview environments.
provide a streamlined process for updating a pull request preview environment, each PR code changes will redeploy the preview environment and also will:

1. provide a real-time build and deployment status updates
2. each entry point will have a seperated preview environment

> 3. ensure a consistent and unique url for each preview environment. The url structure: `https://<gh-repository>-<gh-branch>-<entry-point>-<pr-number>.wing.cloud.app`

Reviewer 2: @ainvoner Instead of having an additional subdomain with the PR number, I'd use a random id mixed with repo information such as the repo name and/or org, like Vercel does. It's easier for us to generate a unique ID this way, and also generate an alias for it using `https://<gh-repository>-<gh-branch>-<entry-point>.wingcloud.app`, if available.

---
title: Wing Cloud Preview Environments
description: This document describes the Wing Cloud Preview Environments feature.
---

# #3696 - Wing Cloud Preview Environments

- **Author(s)**: @ainvoner, @skyrpex
- **Submission Date**: 2023-08-03
- **Stage**: Proposal
- **Stage Date**: 2023-08-03

This document describes the Wing Cloud Preview Environments feature.

## Introduction

The **Wing Cloud Preview Environments** project is designed to achieve two key objectives. Firstly, it aims to offer a preview environment solution for those interested in using winglang.
Secondly, it seeks to enhance the winglang ecosystem and its toolchain by building a production application using the language.

## Requirements

Wing Cloud users can build and deploy Wing applications to an ephemeral (simulated) environment and use the Wing Console UI to interact with it.
Automatically run tests on the preview environment once created and provide a UI for viewing the test results.
Allow developers to collaborate with their teammates using a shared environment.
Seamlessly integrating with GitHub.

### Sign In to Wing Cloud

1. Goto https://wing.cloud and click on sign in.

> 2. If a developer was already signed in before, he will be automatically signed in to Wing Cloud.

Reviewer 3: ```suggestion
2. If a developer was already signed in before, they will be automatically signed in to Wing Cloud.
```

> 2. If a developer was already signed in before, he will be automatically signed in to Wing Cloud.
> 3. If a developer was not signed in before, he will be redirected to GitHub authentication and then redirected back to Wing Cloud.

Reviewer 3: ```suggestion
2. If a developer was already signed in before, they will be automatically signed in to Wing Cloud.
3. If a developer was not signed in before, he will they redirected to GitHub authentication and then redirected back to Wing Cloud.
```

### Authentication and Authorization

1. Wing Cloud Preview Environments will use GitHub for authentication and authorization.
2. Developers will be able to install Wing Cloud Preview Environments on a repository only if they have the GitHub permissions.
3. Developers will be able to create a PR with a preview environment only if they have the necessary GitHub permissions to the repository **and** their GitHub user is signed up to Wing Cloud.
4. Everyone can view a preview environment if they have the url.

### Wing Cloud Preview Environments Developer Experience

> A developer using winglang has the capability to integrate Wing Cloud Preview Environments with their repository.

Reviewer 3: ```suggestion
A developer using winglang has the capability to integrate Wing Cloud Preview Environments with their (github) repository.
```

1. Goto https://wing.cloud and log in
2. Goto https://wing.cloud/preview and Click on "Deploy with Wing Cloud"
3. Complete GitHub application authentication.
4. Grant repository access permission.
5. You are now ready to use Wing Cloud Preview Environments.
6. A welcome email will be sent to you with additional information and links to documentation.

#### Preview Environment for production branch

For the production branch, there is a single preview environment that will constantly be up-to-date with the latest code committed.
A link to the preview environment is available in the repository main page ("About" section).
Production branch environment url structure is: `https://<gh-repository>-<gh-branch>-<entry-point>.wingcloud.app`

#### Preview Environments for Pull Requests

Upon each creation of a pull request an automatic comment will be added to the PR and will guide the developer to dedicated preview environments.
provide a streamlined process for updating a pull request preview environment, each PR code changes will redeploy the preview environment and also will:

1. provide a real-time build and deployment status updates

> 2. each entry point will have a seperated preview environment

Reviewer 4: I think multiple entry points sounds OK to me - if there are too many entry points detected, maybe we can just throw an error (or only create the first 5 etc.). But it also feels like a P2 requirement

> 2. each entry point will have a seperated preview environment

Author: it is a good point, however we agreed to handle it post mvp. see: https://github.com/winglang/wing/pull/3696#discussion_r1285184277

| building.main.w | 🔄 Building ([logs](https://wing.cloud/gh-account/gh-repo/gh-branch/entry-file/logs/))   |                                                                                  | Jul 31, 2023 8:01am  |
| stale.main.w    | ⏸️ Stale ([lean more](https://wing.cloud/gh-account/gh-repo/gh-branch/entry-file/logs/)) |                                                                                  | Jul 31, 2023 8:01am  |

TBD - tests results per entry point

**Disclaimer**: Developers that have write access and can create PRs but didn't signed-up to Wing Cloud will not be able to create preview environments.
In this case the above comment will contain a direct link for signing up to Wing Cloud.
After signing up, the developer will be able to create a preview environment for the PR and one will be created automatically after the next code change.

#### Multiple Entry Points

A developer can configure multiple entry points in a single PR.
Every `*.main.w` file in the repository will be considered as an entry point and will have a dedicated preview environment.

> All preview environments will be available in the PR comment and will be updated upon code changes.

Reviewer 3: What is the experience in the console when a PE is updated? Do you get some popup saying "please wait while redeploying PE..."? Or does it break because the backend container died under the hood?

Reviewer 2: There's no signal of change on the frontend. It'd be a nice feature, though!

The websocket connection will break, so events from backend to frontend will stop working. The rest of the information, which is fetched via HTTP requests, will still work but there will be a drastic change because the environment will be wiped clean (eg, no files in buckets, no events, no logs, etc).

Author: added this as a requirement, @yoav-steinberg good point

#### Populate Environment with Initial Data

Out of scope for this project.

#### Automatically Running Tests

Upon PR creation and code changes, Wing Cloud Preview Environments will automatically run all tests defined for each entry point.
For each test run, a new simulator will be created and will be destroyed upon test completion.

> The test results will be available in the PR comment for each entry point.

Reviewer 3: Might be cool in the console, next to each test you'll automatically have a ✅/❌ based on the test results.
Also what happens if there's a flaky test and the automatic running of tests passes but in the PE it fails? Will it update the comment? Is there a way to mark flaky tests? (Assuming there's some connection between the PE and the test running mechanism).

Reviewer 2: There's an API that we can call to update this comment. I think we can send markdown. And, AFAIK, we can update the comment anytime.

Author: added to the non requirements section in this rfc. we can priorities it later

#### Running Tests Manually

Running tests on your environment is easy using the Wing Console tests UI.
Each test will reload the simulator to create a clean environment.

Once the test is over it will **not** clean up the simulator data, so you can debug the environment and share outputs with your teammates.

#### User-Controlled Environments termination

> Closing a PR will terminate the corresponding preview environments.

Reviewer 3: I really think we'll need some mechanism for suspending stale PE's. If there's important state we might need to store it before suspending. But just spinning up a container per PR means that at some point we'll have lots of containers running forever.

2. The PR's preview environment comment will indicate its inactive status.
3. Changes in code will trigger redeployment for all related preview environments in this PR.

### Analytics and Logs

To improve the product we are collecting data and monitoring the system.

1. We collect preview environments usage analytics only. We won't collect anything about your code and project.
2. OpenTelemetry is used for collecting metrics, logs, traces.
3. Datadog is used for monitoring the system.

### Documentation

Developers reading our docs can easily sign up to Wing Cloud and start using its Preview Environments feature in their repository.
In our docs we have the following information:

1. What is Wing Cloud Preview Environments
2. Getting started instructions
3. How can I configure multiple entry points
4. Populate environment with initial data
5. Running tests

### Security

1. The preview environments shouldn’t be able to access files from other preview environments
2. Only Wing Cloud signed up developers with write access to the repository can create preview environments

## NON Requirements

> 1. Preview environments will require authentication to view.

Reviewer 3: What about non-public github repos? I think if we support them the PE needs to be github authenticated.

Author: for the MVP we decided to do what Vercel is doing, which is not to have authentication for PE by default. In order to have auth in Vercel PE you should specifically enable the feature and pay extra for it.
We will support auth protected envs. post MVP

Reviewer 3: So in that case a private vercel repo has a public preview env? If so maybe we need some mangling of the DNS? What does vercel do? Or perhaps just warn the user that since they signed up to wing cloud and the repo is private they should be aware that we're deploying a public PE.

Reviewer 2: We can, although I'll need to design that. There's, again, security implications we have to handle.

Author: this is how vercel is doing it:
https://vercel.com/docs/concepts/deployments/deployment-protection#vercel-authentication

TL;DR
- Free to all plans:
enable vercel authentication to all deployed environment. users should be authenticated in vercel before viewing an environment.
- paid
password protected. define specific password for each project.

I think we can start with the first option. WDYT?

2. Assign users to Teams.
3. enable preview environments viewers to add comments in the environment itself

## Technical Design

### Technology

> - https://winglang.io for the infrastructure (although it may prove a challenge due to time constrains)

Reviewer 3: Would be interesting to understand the main concerns here

Reviewer 2: I fear hitting a wall but it's being OK so far :)

- AWS for the infrastructure
- Node.js 18 for the backend
- https://github.com/probot/probot for the GitHub integration
- https://fly.io for the preview environments
- https://react.dev/ for the frontend

### Architecture

- A GitHub app
- A Cloud.Function that will be triggered by GitHub events, such as `pull_request.synchronize` or `pull_request.closed`
- A Flyio account to deploy the preview environments
- A Cloud.Table to store different data:
  - The registered users
  - The connected GitHub repositories

>   - The related Flyio projects and environments

Reviewer 3: * some timer/service for cleanups/stale PRs
* some storage for logs

Reviewer 2: Correct, will update the list.

    fn2[Enqueue Events Fn]
    ghapp -- Events --> fn2
    fn2 --> queue
    queue --> fn
    gh -- Integration --> ghapp
    fn --> rest[...]
```

The probot npm package will take care of ensuring the events received are legitimate.

#### Process: Creating Preview Environments

When a GitHub event `pull_request.opened` is received, we will look for `*.main.w` files and create a preview environment for each of them. We will use the Flyio API to create a new project(s) and deploy the code to it (the Flyio specifics may change).

New subdomains will be created for each preview environment:

- `<gh-repository>-<nanoid>.wingcloud.app`
- `<gh-repository>-<gh-branch>-<entry-point>.wingcloud.app`, as an alias of the above, if available. Otherwise, a random ID could be added somewhere, such as in `<gh-repository>-<gh-branch>-<nanoid>-<entry-point>.wingcloud.app`

During this process, we will also create new entries in the Cloud.Table to store the Flyio project ID, the GitHub PR ID, the entry points and whatever unique IDs we generate for every environment. This data is necessary to later resolve the subdomains to the Flyio projects, to update them later, and to delete them.

> The technical details on how the subdomain resolution will work at DNS level are still to be determined.

Reviewer 3: Did we ever discuss a `cloud.Dns` abstraction? Wouldn't this be useful here?

Reviewer 2: I don't recall, but sure it'd be helpful.

---

# #3696 - Wing Cloud Preview Environments

- **Author(s)**: @ainvoner, @skyrpex
- **Submission Date**: 2023-08-03
- **Stage**: Proposal
- **Stage Date**: 2023-08-03

This document describes the Wing Cloud Preview Environments feature.

## Introduction

The **Wing Cloud Preview Environments** project is designed to achieve two key objectives. Firstly, it aims to offer a preview environment solution for those interested in using winglang.
Secondly, it seeks to enhance the winglang ecosystem and its toolchain by building a production application using the language.

## Requirements

Wing Cloud users can build and deploy Wing applications to an ephemeral (simulated) environment and use the Wing Console UI to interact with it.
Automatically run tests on the preview environment once created and provide a UI for viewing the test results.
Allow developers to collaborate with their teammates using a shared environment.
Seamlessly integrating with GitHub.

### Sign Up to Wing Cloud

1. Goto https://wing.cloud and click on sign up.
2. sign up to Wing Cloud using GitHub authentication.

### Sign In to Wing Cloud

1. Goto https://wing.cloud and click on sign in.

> 2. If a developer was already signed in before, he will be automatically signed in to Wing Cloud.

Reviewer 2: @ainvoner signing in and signing up are the same process. I'd simplify it to just "Sign In".

### Analytics and Logs

To improve the product we are collecting data and monitoring the system.

1. We collect preview environments usage analytics only. We won't collect anything about your code and project.
2. OpenTelemetry is used for collecting metrics, logs, traces.
3. Datadog is used for monitoring the system.

### Documentation

Developers reading our docs can easily sign up to Wing Cloud and start using its Preview Environments feature in their repository.
In our docs we have the following information:

1. What is Wing Cloud Preview Environments
2. Getting started instructions
3. How can I configure multiple entry points
4. Populate environment with initial data
5. Running tests

### Security

1. The preview environments shouldn’t be able to access files from other preview environments
2. Only Wing Cloud signed up developers with write access to the repository can create preview environments

## NON Requirements

1. Preview environments will require authentication to view.
2. Assign users to Teams.
3. enable preview environments viewers to add comments in the environment itself

## Technical Design

### Technology

> - https://wing.cloud for the infrastructure (although it may prove a challenge due to time constrains)

Author: @skyrpex did you mean winglang?

Reviewer 2: Yes, thanks.

- AWS for the infrastructure
- Node.js 18 for the backend
- https://github.com/probot/probot for the GitHub integration
- https://fly.io for the preview environments
- https://react.dev/ for the frontend

### Architecture

- A GitHub app
- A Cloud.Function that will be triggered by GitHub events, such as `pull_request.synchronize` or `pull_request.closed`
- A Flyio account to deploy the preview environments
- A Cloud.Table to store different data:
  - The registered users

>   - The Wing Cloud projects and their config

Author: is it relevant for the mvp? (next line as well)

Reviewer 2: Not anymore since we don't want configuration yet. I'll rephrase it.

```mermaid
graph LR
    gh{GitHub}
    ghapp>GitHub App]
    fn[Process Events Fn]
    flyio{Flyio}
    gh -- Integration --> ghapp
    ghapp -- Events --> fn
    fn -- Create Preview Environment --> flyio
    fn -- Update Preview Environment --> flyio
    fn -- Delete Preview Environment --> flyio
```

Depending on whether GitHub retries events or not, we may want to include a FIFO queue in the middle.

```mermaid
graph LR
    gh{GitHub}
    ghapp>GitHub App]
    fn[Process Events Fn]
    queue[(Events Queue)]
    fn2[Enqueue Events Fn]
    ghapp -- Events --> fn2
    fn2 --> queue
    queue --> fn
    gh -- Integration --> ghapp
    fn --> rest[...]
```

The probot npm package will take care of ensuring the events received are legitimate.

#### Process: Creating Preview Environments

When a GitHub event `pull_request.opened` is received, we will look for `*.main.w` files and create a preview environment for each of them. We will use the Flyio API to create a new project(s) and deploy the code to it (the Flyio specifics may change).

> During this process, we will also create new entries in the Cloud.Table to store the Flyio project ID, the GitHub PR ID, and the entry point.

Author: can you please explain why do we need to save this information?
(we discussed about reducing the gh api requests for minimum - this was one reason for saving the PR and the VM ID, for knowing which PR to update if the VM was stale for X amount of time)

Reviewer 2: I rephrased everything a bit, but those reasons are correct. Another reason is that we have to resolve a domain such as `<gh-repository>-<gh-branch>-<entry-point>.wingcloud.app` or `<gh-repository>-<nanoid>.wingcloud.app` to a Flyio domain.

---
title: Wing Cloud Preview Environments
description: This document describes the Wing Cloud Preview Environments feature.
---

# #3696 - Wing Cloud Preview Environments

- **Author(s)**: @ainvoner, @skyrpex
- **Submission Date**: 2023-08-03
- **Stage**: Proposal
- **Stage Date**: 2023-08-03

This document describes the Wing Cloud Preview Environments feature.

## Introduction

The **Wing Cloud Preview Environments** project is designed to achieve two key objectives. Firstly, it aims to offer a preview environment solution for those interested in using winglang.
Secondly, it seeks to enhance the winglang ecosystem and its toolchain by building a production application using the language.

> ## Requirements

Reviewer 2: This implies implementing some basic authorization roles so different people can access/maintain the projects.

> ## Requirements

Reviewer: Yes I think there's no way around this

1. What is Wing Cloud Preview Environments
2. Getting started instructions
3. How can I configure multiple entry points
4. Populate environment with initial data
5. Running tests

### Security

1. The preview environments shouldn’t be able to access files from other preview environments
2. Only Wing Cloud signed up developers with write access to the repository can create preview environments

## NON Requirements

1. Preview environments will require authentication to view.
2. Assign users to Teams.
3. enable preview environments viewers to add comments in the environment itself
4. update the PR comment with the latest test results

## Technical Design

### Technology

- https://winglang.io for the infrastructure
- AWS for the infrastructure
- Node.js 18 for the backend
- https://github.com/probot/probot for the GitHub integration
- https://fly.io for the preview environments
- https://react.dev/ for the frontend

### Architecture

- A GitHub app
- A Cloud.Function that will be triggered by GitHub events, such as `pull_request.synchronize` or `pull_request.closed`
- A Flyio account to deploy the preview environments
- A Cloud.Table to store different data:

>   - The registered users

Author: @skyrpex we should start thinking about how to group users and preview environments.
I've added those fields so we will have the needed data in our DB, and once we will start implementing the environments dashboard we will already have the needed data per user and env.
When thinking about it, we will also need a table to know which env. belongs to which gh project / repo / user.

WDYT?

For the repository production branch, there is a single preview environment that will constantly be up-to-date with the latest code committed.
A link to the preview environment is available in the repository main page ("About" section).
Production branch environment url structure is: `https://<gh-repository>-<gh-branch>-<entry-point>.wingcloud.app`
Once the installation of the Wing Cloud application is completed, an initial preview environment will be created and will be available in the repository main page.

#### Preview Environments for Pull Requests

Upon each creation of a pull request an automatic comment will be added to the PR and will guide the developer to dedicated preview environments.
provide a streamlined process for updating a pull request preview environment, each PR code changes will redeploy the preview environment and will also:

1. Provide a real-time build and deployment status updates
2. While preview environment is being updated, the user that is currently using the environment will be notified and will be able to refresh the page to get the latest changes.
3. Each entry point will have a seperated preview environment

> 4. Ensure a consistent and unique url for each preview environment. The url structure: `https://<gh-repository>-<gh-branch>-<entry-point>.wingcloud.app`

Reviewer 5: Will we have the option to also have a URL per commit like Vercel does?

Will the 'entry-point' be the full path of the file? (will not be unique otherwise)

I think <entry-point>-<gh-branch>-<repo>.wingcloud.app will be more readable. 

For example:
```
entry1-add-new-feat-my-app.wingcloud.app
entry2-add-new-feat-my-app.wingcloud.app
```

vs

```
my-app-add-new-feat-entry1.wingcloud.app
my-app-add-new-feat-entry1.wingcloud.app
```

2. filter repositories that have Wing Cloud application installed.
3. for each repository, get all preview environments.
4. there might be need to store repository -> preview environments mapping in our DB.
5. there is no need to store user -> repositories mapping in our DB, we can get it from GitHub API.

### Processing GitHub Events

A Cloud.Function will listen to GitHub events and will process them. Based on the event, the function will create, update or delete preview environments on Flyio.

```mermaid
graph LR
    gh{GitHub}
    ghapp>GitHub App]
    fn[Process Events Fn]
    flyio{Flyio}
    gh -- Integration --> ghapp
    ghapp -- Events --> fn
    fn -- Create Preview Environment --> flyio
    fn -- Update Preview Environment --> flyio
    fn -- Delete Preview Environment --> flyio
```

Depending on whether GitHub retries events or not, we may want to include a FIFO queue in the middle.

```mermaid
graph LR
    gh{GitHub}
    ghapp>GitHub App]
    fn[Process Events Fn]
    queue[(Events Queue)]
    fn2[Enqueue Events Fn]
    ghapp -- Events --> fn2
    fn2 --> queue
    queue --> fn
    gh -- Integration --> ghapp
    fn --> rest[...]
```

> The probot npm package will take care of ensuring the events received are legitimate.

Reviewer 5: What part of the `github -> wingcloud` flow will be managed by the probot and which part by the wing project?

---

# #3696 - Wing Cloud Preview Environments

- **Author(s)**: @ainvoner, @skyrpex
- **Submission Date**: 2023-08-03
- **Stage**: Proposal
- **Stage Date**: 2023-08-03

This document describes an MVP for the first Wing Cloud product - Wing Cloud Preview Environments.

## Introduction

**Preview Environments** are the first feature of Wing Cloud. It will also be the first production grade product written in Wing language.

## Requirements

Wing Cloud users can build and deploy Wing applications to an ephemeral environment that uses the Wing Simulator platform, and use the Wing Console UI to interact with it.
Once a preview environment is created, tests will be executed automatically and the results will be available in the preview environment dedicated comment in GitHub.
Each preview environment will have a unique URL which enable developers to work together and share with external customers as needed.

Wing Cloud will be integrated with GitHub.

### Wing Cloud User Experience

#### Sign-In to Wing Cloud (#1)

1. On https://wing.cloud website click on sign in.
2. Authenticate using your GitHub account.

> 3. If you were already signed in, you will see the Wing Cloud dashboard (under https://wing.cloud directly).

Reviewer 6: It make more sense to me to have `app.wing.cloud` or something like that.

Reviewer 6: I understand there where already discussion about this, I am fine with what you decide

Reviewer 2: I actually prefer that, too. It's technically simpler and will avoid us doing some black magic with CloudFront.

#### Redirect to sign-in process from a Preview Environment URL (#2)

1. If you try to access a preview environment URL while you are not signed in to Wing Cloud (from the PR comment for example), you will be temporarily redirected to the Wing Cloud sign-in flow.
2. Once singed in, you will be redirected back to the preview environment URL.

#### Wing Cloud Dashboard for Signed-In Users (#3)

1. Once signed in, you will be able to navigate directly to https://wing.cloud/dashboard.
2. You can see and manage all your projects and teams.
3. You will see all preview environments associated with your personal account and teams you are member in.
4. In order to see a preview environment, just click on the preview environment URL from the list of all your preview environments.
5. For each preview environment, you can see the deployment logs and the test results and logs.

#### Create a new personal Project (#4)

1. To create a new project, click on "New Project" under "Personal Account".

> 2. Choose between the GitHub repositories that you have access to.

Reviewer: project:repo == 1:1 ?

> 3. Define your secret values.

Reviewer: environment variables

> 3. Define your secret values.

Reviewer: environment variables

> 4. Make sure there's a `main.w` file in the root of the repository.

Reviewer: ```suggestion
4. Make sure there's a `main.w` or `*.main.w` file in the root of the repository.
```

5. Click on "Deploy".

> 6. You'll be redirected to your new project's page (https://wing.cloud/dashboard/-user/-project).

Reviewer: ```suggestion
6. You'll be redirected to your new project's page (https://wing.cloud/-user/-project).
```

> #### Create a new Team (#5)

Reviewer: are teams p1 for the mvp? can we allow users to add other users to their personal projects without a team?

> 6. You'll be redirected to your new project's page (https://wing.cloud/-user/-project).

Reviewer 6: Personal account is mentioned here and also `-user` in the path. 
Did you guys have the organization vs personal account discussion? 

From my experience this is a complex topic, especially when you start to think about analytics, permission, multi organizations and migrating projects. 

I suggest we spend some time together on this topic to make sure we start as simple as possible.

1. Click on "New Team".
2. Write a team name.
3. Optionally, invite other Wing Cloud users to the team (by GitHub user ID).

> #### Create a new team Project (#6)

Reviewer: seems duplicated with the "personal team"  no need to duplicate

1. Click on "New Project" under the team of your choice.
2. Choose between the GitHub repositories that you have access to.
3. Define your secret values.
4. Make sure there's a `main.w` file in the root of the repository.
5. Click on "Deploy".

> 6. You'll be redirected to your new project's page (https://wing.cloud/-team/-project).

Reviewer 2: Yes, they will be two separate projects that happen to use the same repository as source. Both projects will be governed by different people (one by you, the other one by a team), and will have different secrets.

> #### Preview Environment for production branch (#7)

Reviewer: i would frame this as "preview environment for branches", and then allow users to basically create an environment for any branch in the repo (`main` is the just a case of that)

maybe p2 for mvp

For the repository production branch, there is a single preview environment that will constantly be up-to-date with the latest code committed.
A link to the preview environment is available in the repository main page ("About" section).

> Production branch environment url structure is: `https://wing.cloud/-account/-project/-branch`

Reviewer 6: What is `account` here?

Reviewer 2: Either a team ID or user ID

Once the installation of the Wing Cloud application is completed, an initial preview environment will be created and will be available in the repository main page.

> #### Preview Environments for Pull Requests (#8)

Reviewer: this is p1.

Upon each creation of a pull request an automatic comment will be added to the PR and will guide the developer to a dedicated preview environment.
Provide a streamlined process for updating a pull request preview environment, each code changes in a PR will redeploy the preview environment and will also:

> 1. Provide a real-time build and deployment status updates

Reviewer: what about tests?

2. While preview environment is being updated, the user that is currently using the environment will be notified and will be able to refresh the page to get the latest changes.

> 3. Ensure a consistent and unique url for each preview environment. The url structure: `https://wing.cloud/dashboard/-account/-project/-branch` (account is a <user> or a <team>)

Reviewer: ```suggestion
3. Ensure a consistent and unique url for each preview environment. The url structure: `https://wing.cloud/-account/-project/-branch` (account is a <user> or a <team>)
```

If we really need another URL component, it should be `/environments` or `/projects`

| failed.main.w   | ❌ Failed ([logs](https://wing.cloud/dashboard/-account/-project/-branch/logs/build/))     |                                                               |                                                                                                                                                        | Jul 31, 2023 8:01am |
| building.main.w | 🔄 Building ([logs](https://wing.cloud/dashboard/-account/-project/-branch/logs/build/))   |                                                               |                                                                                                                                                        | Jul 31, 2023 8:01am |
| stale.main.w    | ⏸️ Stale ([lean more](https://wing.cloud/dashboard/-account/-project/-branch/logs/build/)) |                                                               |                                                                                                                                                        | Jul 31, 2023 8:01am |

#### Preview Environment Environment Variables (Secrets) (#9)

> The project management dashboard will include a section to manage the secrets of the project. These secrets will be available during the Wing compilation.

Reviewer: is this at the project level?
how do we handle the case of different environment variables per environment (e.g. prod/staging).
maybe p2

#### Run Tests Automatically Upon PR Creation and Code Changes (#10)

Upon PR creation and code changes, Wing Cloud Preview Environments will automatically run all tests defined for the entry point.

> For each test run, a new simulator will be created and will be destroyed upon test completion.

Reviewer: ```suggestion
For each test run, a new simulator instance will be created and will be destroyed upon test completion.
```

The test results will be available in the PR comment with links to the tests logs (`https://wing.cloud/dashboard/-account/-project/-branch/logs/tests/<test-name>`)

#### Running Tests Manually From Wing Console (#11)

Running tests on your environment is easy using the Wing Console tests UI.
Each test will reload the simulator to create a clean environment.

Once the test is over it will **not** clean up the simulator data, so you can debug the environment and share outputs with your teammates.

#### User-Controlled Environments termination (#12)

Closing a PR will terminate the corresponding preview environments.
The PR comment will indicate the termination of each preview environment and the URLs won't be valid anymore.

#### Self-Cleaning Environments Mechanism (#13)

To ensure efficient resource utilization:

> 1. Preview environments associated with stale PRs (without access or code changes for over 30 days) will automatically deactivate.

Reviewer 7: Perhaps being more aggressive could make sense? Perhaps 3 or 7 days?

2. The PR's preview environment comment will indicate its inactive status.
3. Changes in code will trigger redeployment for all related preview environments in this PR.

> 

Reviewer: missing: how do users access application endpoints, such as websites and api endpoints?

### Analytics and Logs (#14)

To improve the product we are collecting data and monitoring the system.

1. We collect preview environments usage analytics only. We won't collect anything about your code and project.
2. OpenTelemetry is used for collecting metrics, logs, traces.
3. Datadog is used for monitoring the system.

### Documentation (#15)

Developers reading our docs can easily sign up to Wing Cloud and start using its Preview Environments feature in their repository.
In our docs we have the following information:

1. What is Wing Cloud Preview Environments
2. Getting started instructions
3. How can I configure multiple entry points
4. Populate environment with initial data
5. Running tests

### Security (#16)

1. Preview environments shouldn’t be able to access files from other preview environments.
2. Only Wing Cloud signed up developers with access to the repository can view preview environments created for this repository.

## Non-Requirements

1. password protected preview environments.
2. wing configuration file support. (for stating a specific wing version, entry points, etc.)

> 3. update PR comment with tests results triggered from the console.

Reviewer: i don't think this will even be needed

4. enable preview environments viewers to add comments in the environment.
5. populate Environment with Initial Data
6. Allow users to define specific environment machine type / configuration (cpu, memory, etc.)
7. Support multiple entry points in a single PR:

   There is a support for multiple entry points in a single PR.
   Every `*.main.w` file in the repository will be considered as an entry point and will have a dedicated preview environment.
   All preview environments will be available in the PR comment and will be updated upon code changes.

## Technical Details

### Entity Relationship Diagram

```mermaid
erDiagram

>     User ||--0+ RoleAssignment : has

Reviewer: I don't think users should own projects, just teams.

#### Create a new team Project (#6)

1. Click on "New Project" under the team of your choice.
2. Choose between the GitHub repositories that you have access to.
3. Define your secret values.
4. Make sure there's a `*.main.w` file in the root of the repository.
5. Click on "Deploy".
6. You'll be redirected to your new project's page (https://wing.cloud/dashboard/-team/-project).

#### Preview Environment for production branch (#7)

For the repository production branch, there is a single preview environment that will constantly be up-to-date with the latest code committed.
A link to the preview environment is available in the repository main page ("About" section).
Production branch environment url structure is: `https://wing.cloud/-account/dashboard/-project/-branch`
Once the installation of the Wing Cloud application is completed, an initial preview environment will be created and will be available in the repository main page.

#### Preview Environments for Pull Requests (#8)

Upon each creation of a pull request an automatic comment will be added to the PR and will guide the developer to a dedicated preview environment.

> Provide a streamlined process for updating a pull request preview environment, each code changes in a PR will redeploy the preview environment and will also:

Reviewer 8: ```suggestion
Provide a streamlined process for updating a pull request preview environment, each commit in a PR will redeploy the preview environment and will also:
```

1. Provide a real-time build, tests and deployment status updates
2. While preview environment is being updated, the user that is currently using the environment will be notified and will be able to refresh the page to get the latest changes.

> 3. Ensure a consistent and unique url for each preview environment. The url structure: `https://wing.cloud/dashboard/-account/-project/-branch` (account is a <user> or a <team>)

Reviewer 8: `<something>` (without backticks) does not render on markdown, so I added backticks:
```suggestion
3. Ensure a consistent and unique url for each preview environment. The url structure: `https://wing.cloud/dashboard/-account/-project/-branch` (account is a `<user>` or a `<team>`)
```

| building.main.w | 🔄 Building ([logs](https://wing.cloud/dashboard/-account/-project/-branch/logs/build/))   |                                                               |                                                                                                                                                        | Jul 31, 2023 8:01am |
| stale.main.w    | ⏸️ Stale ([lean more](https://wing.cloud/dashboard/-account/-project/-branch/logs/build/)) |                                                               |                                                                                                                                                        | Jul 31, 2023 8:01am |

#### Preview Environment Environment Variables (Secrets) (#9)

The project management dashboard will include a section to manage the secrets of the project. These secrets will be available during the Wing compilation.

> #### Run Tests Automatically Upon PR Creation and Code Changes (#10)

Reviewer 8: ```suggestion
#### Run Tests Automatically Upon PR Creation and Code Commits (#10)
```

> Upon PR creation and code changes, Wing Cloud Preview Environments will automatically run all tests defined for the entry point.

Reviewer 8: ```suggestion
Upon PR creation and code commits, Wing Cloud Preview Environments will automatically run all tests defined for the entry point.
```

For each test run, a new simulator instance will be created and will be destroyed upon test completion.
The test results will be available in the PR comment with links to the tests logs (`https://wing.cloud/dashboard/-account/-project/-branch/logs/tests/<test-name>`)

#### Running Tests Manually From Wing Console (#11)

Running tests on your environment is easy using the Wing Console tests UI.
Each test will reload the simulator to create a clean environment.

Once the test is over it will **not** clean up the simulator data, so you can debug the environment and share outputs with your teammates.

#### Access application endpoints from the Wing Console (#12)

Users will be able to access the applications' public endpoints such as APIs and web pages.

#### User-Controlled Environments termination (#13)

Closing a PR will terminate the corresponding preview environments.
The PR comment will indicate the termination of each preview environment and the URLs won't be valid anymore.

> #### Self-Cleaning Environments Mechanism (#14)

Reviewer 8: ```suggestion
#### Environments Self-Cleaning Mechanism (#14)
```

To ensure efficient resource utilization:

> 1. Preview environments associated with stale PRs (without access or code changes for over 10 days) will automatically deactivate.

Reviewer 8: ```suggestion
1. Preview environments associated with stale PRs (without access or code commits for over 10 days) will automatically deactivate.
```

In our docs we have the following information:

1. What is Wing Cloud Preview Environments
2. Getting started instructions
3. How can I configure multiple entry points
4. Populate environment with initial data
5. Running tests

### Security (#17)

1. Preview environments shouldn’t be able to access files from other preview environments.
2. Only Wing Cloud signed up developers with access to the repository can view preview environments created for this repository.

## Non-Requirements

1. password protected preview environments.
2. wing configuration file support. (for stating a specific wing version, entry points, etc.)
3. update PR comment with tests results triggered from the console.
4. enable preview environments viewers to add comments in the environment.
5. populate Environment with Initial Data
6. Allow users to define specific environment machine type / configuration (cpu, memory, etc.)
7. Support multiple entry points in a single PR:

   There is a support for multiple entry points in a single PR.
   Every `*.main.w` file in the repository will be considered as an entry point and will have a dedicated preview environment.

>    All preview environments will be available in the PR comment and will be updated upon code changes.

Reviewer 8: ```suggestion
   All preview environments will be available in the PR comment and will be updated upon code commit.
```

   - dashboard and user management ui
5. enable preview environments viewers to add comments in the environment.
6. populate Environment with Initial Data
7. Allow users to define specific environment machine type / configuration (cpu, memory, etc.)
8. Support multiple entry points in a single PR:

   There is a support for multiple entry points in a single PR.
   Every `*.main.w` file in the repository will be considered as an entry point and will have a dedicated preview environment.
   All preview environments will be available in the PR comment and will be updated upon code changes.

## Technical Details

### Entity Relationship Diagram

```mermaid
erDiagram
    User ||--0+ RoleAssignment : has
    RoleAssignment 1+--|| Team : has
    User o|--0+ Project : owns
    Team o|--0+ Project : owns
    User {
        UserId userId
        string githubUserId
        string name
        string email
    }
    RoleAssignment {
        UserId userId
        string roleType
    }
    Project {
        ProjectId projectId
        UserIdOrTeamId accountId
        string name
        string githubRepoId
        string wingEntrypoint

>         map environmentVariables

Reviewer 2: Yes, I couldn't figure how to describe such data structure in mermaid. Vercel also does this:

<img width="941" alt="image" src="https://github.com/winglang/wing/assets/1077520/3ed32998-2956-4425-81d7-5dfb04e07550">

---
title: Wing Preview Environments
description: This document describes the Wing Cloud Preview Environments feature.
---

# #3696 - Wing Cloud Preview Environments

- **Author(s):**: @ainvoner, @skyrpex
- **Submission Date**: 2023-08-03
- **Stage**: Proposal
- **Stage Date**: 2023-08-03

This document describes the Wing Cloud Preview Environments feature.

## Introduction
The **Wing Cloud Preview Environments** project is designed to achieve two key objectives. Firstly, it aims to offer a preview environment solution for those interested in using winglang.
Secondly, it seeks to enhance the winglang ecosystem and its toolchain by building a production application using the language.

## Requirements
Build and Deploy wing applications to a Wing Console simulated environment.
Allow developers to iterate with their teammates using a shared Wing Console instance.
Seamlessly integrating with GitHub.

### Developer Experience
A developer using winglang has the capability to integrate preview environments with his/her repository.

#### Installation
The installation process is straightforward using GitHub application:

> 1. Goto https://wing.cloud/preview website and click "Deploy with Wing"

Author: updated this section, let me know if you still thinks we should change it to "Sign up"

2. Complete GitHub authentication.

> 3. Grant repository access permission.

Author: In my mind it should be something similar to Vercel signup as well (see image). What you describe it the flow in the Vercel dashboard which I think we should not handle in this product MVP.

![Screenshot 2023-08-06 at 10 24 26](https://github.com/winglang/wing/assets/2538825/413dfe7f-c288-4795-8747-a8afccd66454)

#### Preview Environment for production branch
For the production branch, there is a single preview environment that will constantly be up to date with the latest code committed.

> A link to the preview environment is available in the repository main page ("About" section).

Author: we can. this is how vercel is doing it:

![Screenshot 2023-08-06 at 10 42 42](https://github.com/winglang/wing/assets/2538825/ecd76926-dacb-4a40-b83f-5f0d27f0e8e7)

#### Preview Environments for Pull Requests
Upon each creation of a pull request, an automatic comment will be added to the PR and will guide the developer to a dedicated preview environment. The deployed preview environment will have a unique url for easy identification.
To provide a streamlined process for updating a pull request preview environment, each PR code changes will redeploy the preview environment and also will:
1. provide a real-time build and deployment updates
2. ensure a consistent url for the preview environment

> 3. provide the ability to download the deployment logs

Reviewer: What about test output?

> 3. provide the ability to download the deployment logs

Author: updated. let me know if it is still not clear

PR comment example:

| Entry Point  | Status | Preview | Updated (UTC) |
| ------------- | ------------- | ------------- | ------------- | 
| main.w  |  ✅ Ready ([logs](https:/main-pr.wing-preview.com/logs)) | [Visit Preview](https:/main-pr.wing-preview.com) |  Jul 31, 2023 8:01am |
| failed.main.w  | ❌ Failed ([logs](https:/main-pr.wing-preview.com/logs))  |  |  Jul 31, 2023 8:01am |
| building.main.w  |  🔄 Building ([logs](https:/main-pr.wing-preview.com/logs))  | |  Jul 31, 2023 8:01am |
| stale.main.w  | ⏸️ Stale  ([lean more](https:/winglang.io/docs/preview))  | |  Jul 31, 2023 8:01am |

#### User-Controlled Environments termination
User should have control over when to terminate their preview environments. Closing a PR will terminate the corresponding preview environment

#### Self-Cleaning Environments Mechanism
To ensure efficient resource utilization:
1. Preview environments associated with stale PRs (without access or code changes for over 30 days) will automatically deactivate.
2. The PR's preview environment comment will indicate its inactive status.
3. Changes in code will trigger redeployment.

#### Additional Features
Other valuable features include:

> 1. Running tests on pull requests, noting that tests will operate within isolated environments to prevent disruption for other users.

Reviewer: From a UX perspective, I think posting the test results in the PR as part of the comment would be freaking amazing.

> 1. Running tests on pull requests, noting that tests will operate within isolated environments to prevent disruption for other users.

Author: added to the rfc, although I think it should be a P2. adding the test results to the PR will add complexity
when we will have time estimations we will be able to decide if you add this feature or not

2. Configuration of multiple entry points files for creating multiple preview environments in a single PR
3. Populate a preview environment with an initial data

> 

Reviewer: Maybe for MVP we don't even need the web UI

> 

Author: I don't think it should be part of the MVP. We should store the user / gh account / repository so we will be able to add this dashboard and functionalities later

### Analytics and Logs
To improve the product we should collect data that we can analyze
1. Collect usage analytics
2. use OpenTelemetry for collecting metrics, logs, traces

> 

Reviewer: What about monitoring?

### Documentation
Developer reading our docs can easily start integrate Wing Preview Environments in his repository.
In our docs we have information about:
1. What is Wing Preview Environment
2. Getting started instructions
4. How can I configure multiple entry points
5. Populate environment with initial data
6. Running tests

### Security

> 1. The preview environments shouldn’t be able to access files from other preview environments

Reviewer 2: I guess flyio got us covered here. If every environment uses a container, it should be already safe by default.

2. Only authenticated users should be able to use Wing Preview Environments

## Technical Design

> TBD

Reviewer 2: @eladcon can we create docker containers in flyio? also, can we open multiple ports to the outside?

> TBD

Reviewer: I think we need to tackle the port proxying issue generally (and not just for preview environments). See #2227
